doc/netplan.md

```diff
@@ -678,6 +678,67 @@ wpasupplicant installed if you let the ``networkd`` renderer handle wifi.
           ``active-backup``, ``balance-alb``, and ``balance-tlb`` modes.
 
 
+## Properties for device type ``tunnels:``
+
+``parent`` (scalar)
+
+:    The parent device for the tunnel (optional).
+
+``mode`` (scalar)
+
+:   Defines the tunnel mode. Valid options are ``sit``, ``gre``, ``ip6gre``,
+    ``ipip``, ``ipip6``, ``ip6ip6``, ``vti``, and ``vti6``. Additionally,
+    the ``networkd`` backend also supports ``gretap`` and ``ip6gretap`` modes.
+    In addition, the ``NetworkManager`` supports ``isatap`` tunnels.
+
+``local`` (scalar)
+
+:   Defines the address of the local endpoint of the tunnel.
+
+``remote`` (scalar)
+
+:   Defines the address of the remote endpoint of the tunnel.
+
+``key``  (scalar) or (mapping)
+:   See ``keys`` below.
+
+``keys`` (scalar) or (mapping)
+
+:   Define keys to use for the tunnel. The key can be a number or a dotted
+    quad (an IPv4 address). It is used for identification of IP transforms.
```

Comment:
I think we should do one of two things here - either make this documentation much more extensive regarding the meaning of keys, which tunnel types accept them, and what they're used for; or (my preference) delegate the explanation of this to the relevant lower-level documentation (perhaps ip-tunnel(8)).  I think it's just confusing to say that a key can be "a number or an IP address" without explaining what this means.

As a point of interest, systemd.netdev(5) claims that key is only used for VTI/VTI6 tunnels, and ip-tunnel claims it's only used for GRE tunnels... you touch on this below, but maybe this means there are some upstream documentation bugs that need sorting.  (Regardless, we should certainly make sure the netplan documentation is correct for the netplan implementation.)

Reply:
I don't understand what you're missing there, aside perhaps from "refer to 'ip-tunnel(8)' for more information". We do explain that they are used to identify IP transforms, and there are not many more ways to put it. Similarly, the expected values are just a number (could specify range, if it was clearer) or a dotted quad.

We're not dealing with documentation issues here: networkd and NetworkManager have different levels of implementation of handling of transform keys, and netplan needs to correctly capture this (I think it does).